File: backend/shared/src/shared/schemas/dto/assignment_template.py
from typing import List, Optional

from pydantic import BaseModel, Field


class AssignmentTemplateEntryDTO(BaseModel):
    shiftId: str = Field(..., description="Shift identifier")
    dayOfWeek: int = Field(
        ..., ge=0, le=6, description="Day of week (0=Monday, 6=Sunday)"
    )
    workerIds: List[str] = Field(
        ..., min_length=1, description="List of worker IDs assigned"
    )


class AssignmentTemplateWeekDataDTO(BaseModel):
    weekNumber: int = Field(..., ge=0, description="Week number (0-based)")
    entries: List[AssignmentTemplateEntryDTO] = Field(
        ..., description="List of assignment entries"
    )


class AssignmentTemplateDTO(BaseModel):
    id: str
    name: str = Field(..., min_length=1, max_length=100)
    teamId: str
    templateType: str = Field(..., pattern="^(standard|even_odd)$")
    weeksData: List[AssignmentTemplateWeekDataDTO]
    description: Optional[str] = Field(None, max_length=500)
    createdBy: str
    createdAt: float
    updatedAt: float


class AssignmentTemplateCreateDTO(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)


class AssignmentTemplateUpdateDTO(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    templateType: Optional[str] = Field(None, pattern="^(standard|even_odd)$")
    weeksData: Optional[List[AssignmentTemplateWeekDataDTO]] = Field(
        None, min_length=1, max_length=8
    )
    description: Optional[str] = Field(None, max_length=500)


class ApplyAssignmentsToTemplateWeekDTO(BaseModel):
    templateId: str = Field(..., description="Template identifier")
    sourceWeekStartDate: float = Field(
        ..., description="Timestamp of source week Monday"
    )
    targetWeekNumber: int = Field(
        ..., ge=0, description="0-based week number in template to update"
    )


class ApplyAssignmentTemplateToDateRangeDTO(BaseModel):
    templateId: str = Field(..., description="Template identifier")
    startDate: float = Field(..., description="Start date timestamp")
    endDate: float = Field(..., description="End date timestamp")
    overwriteExisting: bool = Field(
        default=True, description="Whether to overwrite existing assignments"
    )
    shiftId: Optional[str] = Field(
        None, description="Optional filter: only apply to this shift"
    )


class AssignmentTemplateApplicationResult(BaseModel):
    success: bool = Field(..., description="Whether the operation was successful")
    assignmentsCreated: int = Field(..., description="Number of assignments created")
    assignmentsDeleted: int = Field(..., description="Number of assignments deleted")
    message: str = Field(..., description="Result message")

File: frontend/src/types/assignment-template.ts
import { Dayjs } from 'dayjs';

export enum TemplateType {
  STANDARD = 'standard',
  EVEN_ODD = 'even_odd',
}

export type TemplateViewMode = 'list' | 'view' | 'edit';

export interface AssignmentTemplateEntryDTO {
  shiftId: string;
  dayOfWeek: number;
  workerIds: string[];
}

export interface AssignmentTemplateWeekDataDTO {
  weekNumber: number;
  entries: AssignmentTemplateEntryDTO[];
}

export interface AssignmentTemplateDTO {
  id: string;
  name: string;
  teamId: string;
  templateType: TemplateType;
  weeksData: AssignmentTemplateWeekDataDTO[];
  description?: string;
  createdBy: string;
  createdAt: number;
  updatedAt: number;
}

export interface AssignmentTemplateCreateDTO {
  name: string;
  description?: string;
}

export interface AssignmentTemplateUpdateDTO {
  name?: string;
  description?: string;
  templateType?: TemplateType;
  weeksData?: AssignmentTemplateWeekDataDTO[];
}

export interface ApplyAssignmentsToTemplateWeekDTO {
  templateId: string;
  sourceWeekStartDate: number;
  targetWeekNumber: number;
}

export interface ApplyAssignmentTemplateToDateRangeDTO {
  templateId: string;
  startDate: number;
  endDate: number;
  overwriteExisting: boolean;
  shiftId?: string;
}

export interface AssignmentTemplateApplicationResult {
  success: boolean;
  assignmentsCreated: number;
  assignmentsDeleted: number;
  message: string;
}

export interface AssignmentTemplateListItem {
  id: string;
  name: string;
  description?: string;
  templateType: TemplateType;
  createdBy: string;
  createdAt: Dayjs;
  updatedAt: Dayjs;
  totalEntries: number;
}

export const ASSIGNMENT_TEMPLATE_CONSTRAINTS = {
  MAX_NAME_LENGTH: 100,
  MAX_DESCRIPTION_LENGTH: 500,
  MAX_WEEKS_IN_TEMPLATE: 8,
  MAX_DATE_RANGE_DAYS: 365,
} as const;
